marshal.go

```diff
@@ -0,0 +1,285 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package httprequest
+
+import (
+	"bytes"
+	"encoding"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"reflect"
+	"strings"
+
+	"github.com/julienschmidt/httprouter"
+	"gopkg.in/errgo.v1"
+)
+
+//  Marshal is the counterpart of Unmarshal. It takes information from
+//  x, which must be a pointer to a struct, and returns an HTTP request
+//  using the given method that holds all of the information
+//
+//  The HTTP request will use the given method.  Named fields in the given
+//  baseURL will be filled out from "path"-tagged fields in x to form the
+//  URL path in the returned request.  These are specified as for httprouter.
```

Comment:
// If a field in baseURL is a suffix of the form "*var" (a trailing wildcard element
    // that holds the rest of the path), the marshaled string must begin with a "/".
    // This matches the httprouter convention that it always returns such fields
    // with a "/" prefix.